File: TobogganApp/TobogganApp/ViewModels/PresentationViewModel.swift
//
//  PresentationViewModel.swift
//  TobogganApp
//
//  Created by Igor Laborie on 16/08/2025.
//

import SwiftUI

// MARK: - Connection Status Extension
extension ConnectionStatus {
    var displayText: String {
        switch self {
        case .connecting:
            return "Connecting..."
        case .connected:
            return "Connected"
        case .closed:
            return "Disconnected"
        case .reconnecting:
            return "Reconnecting..."
        case .error:
            return "Connection Error"
        }
    }
    
    var color: Color {
        switch self {
        case .connected:
            return .green
        case .connecting, .reconnecting:
            return .orange
        default:
            return .red
        }
    }
}

// Notification handler implementation
final class NotificationHandler: ClientNotificationHandler, @unchecked Sendable {
    weak var viewModel: PresentationViewModel?
    
    init(viewModel: PresentationViewModel?) {
        self.viewModel = viewModel
    }
    
    func onStateChange(state: State) {
        viewModel?.handleStateChange(state)
    }

    func onTalkChange(state: State) {
        viewModel?.handleTalkChange(state)
    }

    func onConnectionStatusChange(status: ConnectionStatus) {
        viewModel?.handleConnectionStatusChange(status)
    }

    func onRegistered(clientId: String) {
        // Client registration notification - no UI action needed
    }

    func onClientConnected(clientId: String, name: String) {
        // Another client connected - no UI action needed
    }

    func onClientDisconnected(clientId: String, name: String) {
        // Another client disconnected - no UI action needed
    }

    func onError(error: String) {
        viewModel?.handleError(error)
    }
}

// Shared configuration for TobogganClient
enum TobogganConfig {
    static let shared = ClientConfig(
        url: "http://127.0.0.1:8080",
        maxRetries: 3,
        retryDelay: 1.0
    )
}

// ViewModel following KISS principle with TobogganCore integration
class PresentationViewModel: ObservableObject {
    @Published var presentationTitle = "Presentation title - Date"
    @Published var nextSlideTitle = "Next Title"
    @Published var connectionStatus: ConnectionStatus = .closed
    @Published var currentSlideIndex: Int?
    @Published var totalSlides: Int = 0
    @Published var currentSlide: Slide?
    @Published var canGoPrevious = false
    @Published var canGoNext = false

    // Step tracking
    @Published var currentStep: Int = 0
    @Published var stepCount: Int = 0

    // Duration tracking
    @Published var totalDuration: TimeInterval = 0

    var formattedDuration: String {
        let minutes = Int(totalDuration) / 60
        let seconds = Int(totalDuration) % 60
        return String(format: "%02d:%02d", minutes, seconds)
    }

    // Step navigation computed properties
    var canGoPreviousStep: Bool {
        currentStep > 0
    }

    var canGoNextStep: Bool {
        stepCount > 0 && currentStep < stepCount - 1
    }
    
    // Navigation button logic
    // When on first step, previous button becomes "previous slide"
    var showPreviousSlideInsteadOfStep: Bool {
        currentStep == 0
    }
    
    // When on last step, next button becomes "next slide"
    var showNextSlideInsteadOfStep: Bool {
        stepCount == 0 || currentStep >= stepCount - 1
    }

    // Error dialog state
    @Published var showErrorAlert = false
    @Published var errorMessage = ""

    private let tobogganClient: TobogganClient
    private var currentState: State?
    private let notificationHandler: NotificationHandler
    private var pendingStateUpdate: State?
    private var talkLoaded = false
    
    init() {
        // Initialize notification handler and client synchronously
        self.notificationHandler = NotificationHandler(viewModel: nil)
        self.tobogganClient = TobogganClient(
            config: TobogganConfig.shared,
            clientName: "TobogganApp",
            handler: self.notificationHandler
        )
        self.notificationHandler.viewModel = self
        
        // Connect asynchronously
        connectAndFetchTalk()
    }
    
    // MARK: - TobogganCore Integration
    
    private func connectAndFetchTalk() {
        // Connect and fetch talk info on background thread
        DispatchQueue.global(qos: .background).async { [weak self] in
            guard let self = self else { return }
            
            // Update connection status
            Task { @MainActor in
                self.connectionStatus = .connecting
            }
            
            // Connect to server
            self.tobogganClient.connect()
            
            self.fetchTalkInfoDirect(client: self.tobogganClient)
        }
    }
    
    private func fetchTalkInfoDirect(client: TobogganClient) {
        if let talk = client.getTalk() {
            // Note: talk.slides contains slide IDs (which happen to be titles in current implementation)
            // We don't store them - we'll fetch slides on demand using client.getSlide()
            
            Task { @MainActor in
                // Update presentation title
                presentationTitle = "\(talk.title) - \(talk.date)"
                
                // Store the count of slides
                totalSlides = talk.slides.count
                
                // Mark talk as loaded
                talkLoaded = true
                
                // Process any pending state updates
                if let pendingState = pendingStateUpdate {
                    handleStateChange(pendingState)
                    pendingStateUpdate = nil
                }
            }
        } else {
            Task { @MainActor in
                handleError("Could not fetch talk information from server")
            }
        }
    }
    
    // MARK: - Notification Handlers
    
    func handleStateChange(_ state: State) {
        Task { @MainActor in
            currentState = state

            switch state {
            case .`init`(let totalSlides):
                // In init state, we don't have a current slide yet
                self.totalSlides = Int(totalSlides)
                self.currentStep = 0
                self.stepCount = 0
                updatePresentationState(currentSlideIndex: nil)
            case let .running(previous, current, next, currentStep, stepCount):
                self.currentStep = Int(currentStep)
                self.stepCount = Int(stepCount)
                updatePresentationState(
                    currentSlideIndex: Int(current),
                    previousSlideIndex: previous.map(Int.init),
                    nextSlideIndex: next.map(Int.init)
                )
            case let .done(previous, current, currentStep, stepCount):
                self.currentStep = Int(currentStep)
                self.stepCount = Int(stepCount)
                updatePresentationState(
                    currentSlideIndex: Int(current),
                    previousSlideIndex: previous.map(Int.init)
                )
            }
        }
    }

    func handleTalkChange(_ state: State) {
        print("📝 Presentation updated - refetching talk metadata and slides")

        // Refetch talk information in background
        DispatchQueue.global(qos: .background).async { [weak self] in
            guard let self = self else { return }
            self.fetchTalkInfoDirect(client: self.tobogganClient)
        }

        // Update state to reflect new slide position (server already adjusted)
        handleStateChange(state)
    }
    
    private func updatePresentationState(
        currentSlideIndex: Int?,
        previousSlideIndex: Int? = nil,
        nextSlideIndex: Int? = nil
    ) {
        self.canGoPrevious = (previousSlideIndex != nil)
        self.canGoNext = (nextSlideIndex != nil)
        
        // Update current slide (if we have one)
        if let currentIdx = currentSlideIndex {
            updateSlideFromState(slideIndex: currentIdx)
        } else {
            // In init state - no current slide yet
            currentSlide = nil
            self.currentSlideIndex = nil
        }
        
        // Update next slide title by fetching it on demand
        if let nextIdx = nextSlideIndex,
           let nextSlide = tobogganClient.getSlide(index: UInt32(nextIdx)) {
            nextSlideTitle = nextSlide.title
        } else {
            nextSlideTitle = "<End of presentation>"
        }
    }
    
    func handleConnectionStatusChange(_ status: ConnectionStatus) {
        Task { @MainActor in
            switch status {
            case .connecting:
                connectionStatus = .connecting
            case .connected:
                connectionStatus = .connected
            case .closed:
                connectionStatus = .closed
            case .reconnecting:
                connectionStatus = .reconnecting
            case .error:
                connectionStatus = .error
            }
        }
    }
    
    func handleError(_ error: String) {
        Task { @MainActor in
            connectionStatus = .error
            errorMessage = error
            showErrorAlert = true
        }
    }
    
    private func updateSlideFromState(slideIndex: Int) {
        // Check if talk info has been loaded
        if !talkLoaded {
            pendingStateUpdate = currentState
            return
        }
        
        // Fetch the slide on demand using tobogganClient
        if let slide = tobogganClient.getSlide(index: UInt32(slideIndex)) {
            currentSlide = slide
            
            // Set the slide index directly
            self.currentSlideIndex = slideIndex
        } else {
            Task { @MainActor in
                handleError("Could not fetch slide with index '\(slideIndex)'")
            }
        }
    }
    
    // MARK: - Actions
    
    // MARK: - Step Navigation

    func nextStep() {
        tobogganClient.sendCommand(command: .nextStep)
    }

    func previousStep() {
        tobogganClient.sendCommand(command: .previousStep)
    }

    // MARK: - Slide Navigation

    func nextSlide() {
        tobogganClient.sendCommand(command: .next)
        // Update local index optimistically
        if let current = currentSlideIndex, current < totalSlides - 1 {
            currentSlideIndex = current + 1
        } else if currentSlideIndex == nil {
            // If we're in init state and going next, we start at first slide
            currentSlideIndex = 0
        }
    }

    func previousSlide() {
        tobogganClient.sendCommand(command: .previous)
    }
    
    func firstSlide() {
        tobogganClient.sendCommand(command: .first)
    }
    
    func lastSlide() {
        tobogganClient.sendCommand(command: .last)
    }

    func blink() {
        tobogganClient.sendCommand(command: .blink)
    }
    
    deinit {
        // Client will be cleaned up automatically when deallocated
    }
}

File: toboggan-android/app/src/main/java/com/toboggan/app/viewmodel/PresentationViewModel.kt
package com.toboggan.app.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.Duration
import uniffi.toboggan.ClientConfig
import uniffi.toboggan.ClientNotificationHandler
import uniffi.toboggan.Command
import uniffi.toboggan.ConnectionStatus
import uniffi.toboggan.ErrorKind
import uniffi.toboggan.Slide
import uniffi.toboggan.ClientRole
import uniffi.toboggan.PresentationState
import uniffi.toboggan.TobogganClient

data class PresentationUiState(
    val presentationTitle: String = "Presentation title - Date",
    val connectionStatus: ConnectionStatus = ConnectionStatus.Closed,
    // Audience until the server says otherwise, mirroring the core default: a
    // role that has not arrived is the one that can do the least. Held as a
    // non-null value so no branch can read "unknown" as "allowed".
    val role: ClientRole = ClientRole.AUDIENCE,
    val isRegistered: Boolean = false,
    val currentSlideIndex: Int? = null,
    val previousSlideIndex: Int? = null,
    val nextSlideIndex: Int? = null,
    val totalSlides: Int = 0,
    val currentSlide: Slide? = null,
    val nextSlideTitle: String = "<End of presentation>",
    val errorMessage: String? = null,
    // A refusal is a permissions answer, not a broken connection, and belongs
    // beside the controls rather than in the error banner.
    val notice: String? = null,
    val currentStep: Int = 0,
    // Dots to draw: one per state the slide can be in, which is one more than
    // the number of reveals. Zero when the server has not counted them.
    val stepStates: Int = 0
) {
    val slideProgress: String
        get() = when (currentSlideIndex) {
            null -> "Ready to Start"
            else -> "${currentSlideIndex + 1} of $totalSlides"
        }

    val currentSlideTitle: String
        get() = currentSlide?.title ?: "Ready to Start"

    val isPresenter: Boolean
        get() = role == ClientRole.PRESENTER

    // Derived rather than stored, so a role arriving after a state change still
    // reaches the buttons. Stored, they were computed once from slide adjacency
    // alone and stayed enabled for a client the server would refuse.
    val canGoPrevious: Boolean
        get() = isPresenter && previousSlideIndex != null

    val canGoNext: Boolean
        get() = isPresenter && nextSlideIndex != null
}

class PresentationViewModel : ViewModel(), ClientNotificationHandler {

    private val _uiState = MutableStateFlow(PresentationUiState())
    val uiState: StateFlow<PresentationUiState> = _uiState.asStateFlow()

    private var tobogganClient: TobogganClient? = null
    private var currentState: PresentationState? = null
    private var talkLoaded = false
    private var pendingStateUpdate: PresentationState? = null

    init {
        connectToServer()
    }

    private fun connectToServer() {
        viewModelScope.launch(Dispatchers.IO) {
            val config = ClientConfig(
                // Use 10.0.2.2 for Android emulator localhost
                url = "http://10.0.2.2:8080",
                maxRetries = 3u,
                retryDelay = Duration.ofSeconds(1)
            )

            _uiState.update { it.copy(connectionStatus = ConnectionStatus.Connecting) }

            tobogganClient = TobogganClient(config, "Android Remote", this@PresentationViewModel)
            tobogganClient?.connect()

            fetchTalkInfo()
        }
    }

    private fun fetchTalkInfo() {
        tobogganClient?.getTalk()?.let { talk ->
            viewModelScope.launch(Dispatchers.Main) {
                _uiState.update { state ->
                    state.copy(
                        presentationTitle = "${talk.title} - ${talk.date}",
                        totalSlides = talk.titles.size
                    )
                }
                talkLoaded = true

                // Process any pending state updates
                pendingStateUpdate?.let { pending ->
                    handleStateChange(pending)
                    pendingStateUpdate = null
                }
            }
        } ?: run {
            viewModelScope.launch(Dispatchers.Main) {
                handleError(ErrorKind.TRANSPORT, "Could not fetch talk information from server")
            }
        }
    }

    // MARK: - ClientNotificationHandler implementation

    override fun onStateChange(state: PresentationState) {
        viewModelScope.launch(Dispatchers.Main) {
            handleStateChange(state)
        }
    }

    override fun onTalkChange(state: PresentationState) {
        viewModelScope.launch(Dispatchers.IO) {
            fetchTalkInfo()
        }
        viewModelScope.launch(Dispatchers.Main) {
            handleStateChange(state)
        }
    }

    override fun onConnectionStatusChange(status: ConnectionStatus) {
        viewModelScope.launch(Dispatchers.Main) {
            _uiState.update { it.copy(connectionStatus = status) }
        }
    }

    override fun onError(kind: ErrorKind, error: String) {
        viewModelScope.launch(Dispatchers.Main) {
            handleError(kind, error)
        }
    }

    override fun onRegistered(clientId: String, role: ClientRole) {
        // Said plainly, because the alternative is discovering it by pressing a
        // button and having the server refuse: a client that is not on the
        // presenting machine and carries no token is audience. The role gates
        // `canGoPrevious`/`canGoNext`, and the notice says why they are dead.
        _uiState.update {
            it.copy(
                role = role,
                isRegistered = true,
                notice = if (role == ClientRole.AUDIENCE) {
                    "Watching — this client cannot present."
                } else {
                    null
                }
            )
        }
    }

    override fun onClientConnected(clientId: String, name: String) {
        // Another client connected - no UI action needed
    }

    override fun onClientDisconnected(clientId: String, name: String) {
        // Another client disconnected - no UI action needed
    }

    // MARK: - State handling

    private fun handleStateChange(state: PresentationState) {
        currentState = state

        when (state) {
            is PresentationState.Init -> {
                _uiState.update { it.copy(totalSlides = state.totalSlides.toInt()) }
                updatePresentationState(currentSlideIndex = null)
            }
            is PresentationState.Running -> {
                updatePresentationState(
                    currentSlideIndex = state.current.toInt(),
                    previousSlideIndex = state.previous?.toInt(),
                    nextSlideIndex = state.next?.toInt(),
                    currentStep = state.currentStep.toInt(),
                    stepStates = stepStates(state.stepCount)
                )
            }
            is PresentationState.Done -> {
                updatePresentationState(
                    currentSlideIndex = state.current.toInt(),
                    previousSlideIndex = state.previous?.toInt(),
                    nextSlideIndex = null,
                    currentStep = state.currentStep.toInt(),
                    stepStates = stepStates(state.stepCount)
                )
            }
        }
    }

    // How many dots the slide is worth.
    // 
    // The server counts *additional* reveals, so a slide with two of them has
    // three states. `null` means it has not counted at all, which is not the
    // same as counting zero — see `Slide.stepCount` on the Rust side.
    private fun stepStates(revealCount: UInt?): Int =
        revealCount?.let { it.toInt() + 1 } ?: 0

    private fun updatePresentationState(
        currentSlideIndex: Int?,
        previousSlideIndex: Int? = null,
        nextSlideIndex: Int? = null,
        currentStep: Int = 0,
        stepStates: Int = 0
    ) {
        // Update current slide if we have one
        val currentSlide = currentSlideIndex?.let { idx ->
            if (!talkLoaded) {
                pendingStateUpdate = currentState
                null
            } else {
                tobogganClient?.getSlide(idx.toUInt())
            }
        }

        // Fetch next slide title
        val nextSlideTitle = nextSlideIndex?.let { idx ->
            tobogganClient?.getSlide(idx.toUInt())?.title
        } ?: "<End of presentation>"

        _uiState.update { state ->
            state.copy(
                currentSlideIndex = currentSlideIndex,
                previousSlideIndex = previousSlideIndex,
                nextSlideIndex = nextSlideIndex,
                currentSlide = currentSlide,
                nextSlideTitle = nextSlideTitle,
                currentStep = currentStep,
                stepStates = stepStates
            )
        }
    }

    private fun handleError(kind: ErrorKind, error: String) {
        _uiState.update { state ->
            when (kind) {
                // The server answered and declined. Reporting that as a
                // connection error blamed the network for a permissions
                // decision — and overwrote the status of a socket that is fine.
                ErrorKind.SERVER -> state.copy(notice = error)
                ErrorKind.TRANSPORT -> state.copy(
                    connectionStatus = ConnectionStatus.Error(error),
                    errorMessage = error
                )
            }
        }
    }

    fun clearError() {
        _uiState.update { it.copy(errorMessage = null) }
    }

    // MARK: - Actions

    fun nextStep() {
        tobogganClient?.sendCommand(Command.NextStep)
    }

    fun previousStep() {
        tobogganClient?.sendCommand(Command.PreviousStep)
    }

    fun firstSlide() {
        tobogganClient?.sendCommand(Command.First)
    }

    fun lastSlide() {
        tobogganClient?.sendCommand(Command.Last)
    }

    fun blink() {
        tobogganClient?.sendCommand(Command.Blink)
    }
}
